DocsCode/NavigationTutorial/NavigationTutorialViewController.m

```diff
@@ -0,0 +1,146 @@
+    // Add a gesture recognizer to the map view
+    UITapGestureRecognizer *setDestination = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(didLongPress:)];
+    [self.mapView addGestureRecognizer:setDestination];
+    // #-end-code-snippet: navigation gesture-recognizer-objc
+}
+
+// #-code-snippet: navigation allow-callouts-objc
+// Implement the delegate method that allows annotations to show callouts when tapped
+-(BOOL)mapView:(MGLMapView *)mapView annotationCanShowCallout:(id<MGLAnnotation>)annotation {
+    return true;
+}
+// #-end-code-snippet: navigation allow-callouts-objc
+
+// #-code-snippet: navigation long-press-objc
+-(void)didLongPress:(UITapGestureRecognizer *)sender {
+    if (sender.state != UIGestureRecognizerStateEnded) {
+        return;
+    }
+    
+    // Converts point where user did a long press to map coordinates
+    CGPoint point = [sender locationInView:self.mapView];
+    CLLocationCoordinate2D coordinate = [self.mapView convertPoint:point toCoordinateFromView:self.mapView];
+    
+    // Create a basic point annotation and add it to the map
+    MGLPointAnnotation *annotation = [MGLPointAnnotation alloc];
+    annotation.coordinate = coordinate;
+    annotation.title = @"Start navigtation";
+    [self.mapView addAnnotation:annotation];
+    
+    // Calcuate the route from the user's location to the set destination
+    [self calculateRoutefromOrigin:self.mapView.userLocation.coordinate
+                     toDestination:annotation.coordinate
+                        completion:^(MBRoute * _Nullable route, NSError * _Nullable error) {
+                            if (error != nil) {
+                                printf("Error calculating route");
+                            }
+                        }];
+}
+// #-end-code-snippet: navigation long-press-objc
+
+// #-code-snippet: navigation calculate-route-objc
+-(void)calculateRoutefromOrigin:(CLLocationCoordinate2D)origin
+                 toDestination :(CLLocationCoordinate2D)destination
+                     completion:(void(^)(MBRoute *_Nullable route, NSError *_Nullable error))completion {
+    
+    MBWaypoint *originWaypoint = [[MBWaypoint alloc] initWithCoordinate:origin coordinateAccuracy:-1 name:@"Start"];
+    
+    MBWaypoint *destinationWaypoint = [[MBWaypoint alloc] initWithCoordinate:destination coordinateAccuracy:-1 name:@"Finish"];
+    
+    // Specify that the route is intented for automobiles avoiding traffic
+    MBNavigationRouteOptions *options = [[MBNavigationRouteOptions alloc] initWithWaypoints:@[originWaypoint, destinationWaypoint] profileIdentifier:MBDirectionsProfileIdentifierAutomobileAvoidingTraffic];
+    
+    // Generate the route object and draw it on the map
+    NSURLSessionDataTask *task = [[MBDirections sharedDirections] calculateDirectionsWithOptions:options completionHandler:^(NSArray<MBWaypoint *> * _Nullable waypoints, NSArray<MBRoute *> * _Nullable routes, NSError * _Nullable error){
```

Comment:
This is a very long line — can it be broken into something easier to read? The `_Nullable` annotations could be removed, though maybe they’re also useful for a first-timer.